### PR TITLE
Add check when mounting an S3 bucket at root to see if bucket exists

### LINKED ISSUE #9948: Alluxio allows mounting S3 buckets which don't exist
**Alluxio Version:**
Master branch

**Describe the bug**
Issuing a mount command on an S3 bucket that doesn't exist succeeds.

**To Reproduce**
Mount an S3 bucket that doesn't exist at the root of the S3 bucket.

**Expected behavior**
Some kind of warning/exception when mounting that the bucket is not accessible or doesn't exist.

**Urgency**
High. Users may make simple typos in mounting a bucket and will not be able to tell without further involved investigation that the mount itself is incorrect.

### PR DESCRIPTION
**Pull request by ns1123**
Fixes #9948

### CODE REVIEW

**Comment by gpang** [comment]
@ns1123 why is this check not enough? https://github.com/Alluxio/alluxio/blob/master/core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java#L2642

**Comment by ns1123** [comment]
@gpang the isDirectory() call always returns true if the URI isRoot(). However, the isRoot() check doesn't check if the user can actually access the root of the bucket.

**Comment by gpang** [comment]
@ns1123 I see. Then shouldn't we change the object store implementation of `isDirectory` to fix this? It works as expected for other UFS implementations.

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java [hunk 1]
```diff
@@ -572,6 +573,32 @@ protected String getRootKey() {
     }
   }
 
+  @Override
+  public boolean isDirectory(String path) throws IOException {
```
Comment (gpang) [inline underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: why can't we do the logic in `ObjectUnderFileSystem` so all object stores are fixed?

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java [hunk 2]
```diff
@@ -594,6 +598,20 @@ public boolean isObjectStorage() {
...
+   * Checks if the root of the ObjectUnderFileSystem is listable. This is done once and cached
+   * per UFS to reduce time spent checking if the root is accessible.
+   *
+   * @return whether the root of the UFS is listable and return false if any exception is thrown
+   */
+  public boolean isRootListable() {
+    try {
+      return getObjectListingChunkForPath("", true) != null;
+    } catch (Exception e) {
+      return false;
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java]: Log an error here?

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java [hunk 3]
```diff
@@ -594,6 +598,21 @@ public boolean isObjectStorage() {
...
+   * Checks if the root of the ObjectUnderFileSystem is listable. This is done once and cached
+   * per UFS to reduce time spent checking if the root is accessible.
+   *
+   * @return whether the root of the UFS is listable and return false if any exception is thrown
+   */
+  public boolean isRootListable() {
+    try {
+      return getObjectListingChunkForPath("", true) != null;
+    } catch (Exception e) {
+      LOG.debug("Unable to list bucket root:", e);
```
Comment (gpang) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java]: Is there any more info you can add to this log message? Like the bucket name or anything else?

Comment (ns1123) [reply core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java]: done

File: core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java [hunk 4]
```diff
@@ -594,6 +598,21 @@ public boolean isObjectStorage() {
...
+   * Checks if the root of the ObjectUnderFileSystem is listable. This is done once and cached
+   * per UFS to reduce time spent checking if the root is accessible.
+   *
+   * @return whether the root of the UFS is listable and return false if any exception is thrown
+   */
+  public boolean isRootListable() {
+    try {
+      return getObjectListingChunkForPath("", true) != null;
+    } catch (Exception e) {
+      LOG.debug("Unable to list root of bucket {}:", super.mUri.toString(), e.getMessage());
```
Comment (gpang) [inline core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java]: Is `e.getMessage()` used in the message? And, do we want the full stack trace of the exception, or do we just want the message?

Comment (ns1123) [reply core/common/src/main/java/alluxio/underfs/ObjectUnderFileSystem.java]: Good catch. There is no message for this and we do want the full exception.

**Comment by ns1123** [comment]
Jenkins, test this please

File: underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java (hunk 1 above)
Comment (ns1123) [reply underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java]: done

**Comment by ns1123** [comment]
Jenkins, test this please

**Comment by calvinjia** [comment]
alluxio-bot, merge this please.

**Pull request merged**